tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -1035,8 +1044,30 @@ LogicalResult ConvertTFLAveragePool2DOp::matchAndRewrite(
       return failure();
   }
 
-  rewriter.replaceOpWithNewOp<tosa::AvgPool2dOp>(
-      op, output_type, tfl_avgpool_op.input(), kernel_size, stride, pad);
+  if (input_type.getElementType().isa<UniformQuantizedType>()) {
+    // Search for attribute annotated by --tosa-convert-tfl-uint8 pass.
+    // This is needed in average pool2d to match TFLite rounding behavior.
+    IntegerAttr override_zeropoint_attr =
```

Comment:
These kind of cross-pass action at a distance attribute dependencies need to be avoided. Why can't you have a helper which performs the check in the pattern which sets this attribute and just call it here (i.e. a static `calculateZeropointOverride(TFL::AveragePool2DOp op)`).

Even better you could define the above to return an `IntegerAttr` conditionally if the input_type is quantized (or nullptr otherwise), then collapse all of the branching in this transform.

Reply:
I agree this kind of cross-pass design is bad, but that was the only way I thought I can achieve what I want.

Briefly explain what I'm doing here:
We're trying to match the rounding behavior to TFL::AveragePool2DOp, which has different rounding behavior between QI8 and QU8 (since it's rounding on its storage type/range). But in the pass pipeline, the convert_tfl_uint8 pass converts all the QU8 into QI8, so the pass after it, e.g. legalize_tf and legalize_tfl, can't distinguish between those two cases. And that's why you see the override_zeropoint attribute is set in convert_tfl_uint8.

The trick we played here (which is bad I agree) is to annotate the zeropoint attribute based on if it's QU8 or QI8 before QU8 is converted. And in legalize_tfl here, it checks if such attribute exists. If it does, then we override that zeropoint with what's stored in the attribute, and we build TOSA::AveragePool2dOp with it.

Is the helper function you mentioned something that could solve the problem without using the cross-pass design?

Reply:
Explained in the comment on top. QU8 and QI8 is not distinguishable after convert_tfl_uint8 pass.

Reply:
This piece appears to be the largest blocker for this update. We're aware - as Kevin mentioned - that the approach isn't ideal. Is it ok if we get all other changes done, then do a separate update with a cleaner implementation of this cross pass issue ?